### PR TITLE
VolPy Update

### PR DESCRIPTION
**Pull request by caichangjia**
# Description

I updated VolPy with the following changes:

VolPy:
Added simple thresholding method; added subthreshold extraction
Added utils.py file in volpy folder. It provided functions for visualization, fast manual annotation, computing summary images and running Mask R-CNN
Cleaned up codes; renamed variables for clarity; added documentation 

summary_images.py:
Added functions for computing mean image in parallel; gaussian blur option for correlation image

caiman/estimates.py:  
Fixed a minor bug

# Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Changes in documentation or new examples for demos and/or use cases.

# Branching

- Pull requests with non breaking bug fixes should be made against the **master** branch.
- All other PRs should be made against the **dev** branch.

# Has your PR been tested?

If you're fixing a bug or introducing a new feature it is recommended you run the tests by typing

```python caimanmanager.py test```

and

```python caimanmanager.py demotest```

prior to submitting your pull request. 

Please describe any additional tests that you ran to verify your changes. If they are fast you can also
include them in the folder 'caiman/tests/` and name them `test_***.py` so they can be included in our lists of tests.

### CODE REVIEW

File: caiman/source_extraction/volpy/README.MD [hunk 1]
```diff
@@ -5,3 +5,5 @@ Please refer to that repository for issues regarding this code.
 
 The code contained in this (and only this) folder is distributed under the MIT License. The rest of package is distributed
 under the GNU-GPL v2 license.
+
```
Comment (pgunn) [inline caiman/source_extraction/volpy/README.MD:8]: How did this file end up with a capital extension?

File: caiman/source_extraction/volpy/atm.py [hunk 2]
```diff
@@ -374,6 +375,8 @@ def test_spikeshape(time, tcourse, tcourse_med, tcourse_std, regress_matrix):
         # iteratively remove too small spikes 'slowly' increasing threshold
         for thresh in threshs:
             while np.sum(spikesizes < thresh):
+                import pdb
```
Comment (pgunn) [inline caiman/source_extraction/volpy/atm.py:378]: Consider importing this at the head of the file rather than dynamically. Also, did you mean to leave this in?

File: caiman/source_extraction/volpy/utils.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
+#!/usr/bin/env python3
```
Comment (pgunn) [inline caiman/source_extraction/volpy/utils.py:1]: Conda calls its python "python" rather than "python3". It's probably best to adjust the hashbang

File: caiman/source_extraction/volpy/utils.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+This file create functions used for demo_pipeline_voltage_imaging
+@author: caichangjia
+"""
+#%% 
+import os
+from IPython import get_ipython
+import matplotlib.pyplot as plt
+from matplotlib.widgets import Slider
+import numpy as np
+import tensorflow as tf
```
Comment (pgunn) [inline caiman/source_extraction/volpy/utils.py:14]: Can you alphabetise the external imports by libraryname?

File: caiman/source_extraction/volpy/utils.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+            
+        roughness: int
+            roughness of the selection surface
+            
+    Return:
+        ROIs: 3-D array
+            region of interests 
+            (# of components * # of pixels in x dim * # of pixels in y dim)
+    """
+    get_ipython().run_line_magic('matplotlib', 'auto')
```
Comment (pgunn) [inline caiman/source_extraction/volpy/utils.py:59]: would this work with regular python?

File: caiman/source_extraction/volpy/utils.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+        display_result: boolean
+            if True, the function will plot the result of inference
+        
+    Return:
+        ROIs: 3-D array
+            region of interests 
+            (# of components * # of pixels in x dim * # of pixels in y dim)
+    """
+    from caiman.source_extraction.volpy.mrcnn import visualize, neurons
+    import caiman.source_extraction.volpy.mrcnn.model as modellib
```
Comment (pgunn) [inline caiman/source_extraction/volpy/utils.py:120]: Consider doing these imports at the head of the file rather than dynamically

File: caiman/source_extraction/volpy/utils.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+            ax2.text(0.1, 0.04,'locality_test:'+str(estimates['locality'][idx][i]), horizontalalignment='center', verticalalignment='center', transform = ax2.transAxes)            
+            
+            ax3.cla()
+            ax3.imshow(img, interpolation='None', cmap=plt.cm.gray, vmax=vmax)
+            imgtmp2 = imgtmp.copy()
+            imgtmp2[imgtmp2 == 0] = np.nan
+            ax3.imshow(imgtmp2, interpolation='None',
+                       alpha=0.5, cmap=plt.cm.hot)
+            ax3.axis('off')
+            print('snr:{0}'.format(estimates['snr'][idx][i]))
```
Comment (pgunn) [inline caiman/source_extraction/volpy/utils.py:268]: consider using f-strings

File: caiman/source_extraction/volpy/utils.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+    fig.canvas.mpl_connect('key_release_event', arrow_key_image_control)
+    plt.show()
+
+
+
+
+
+
+
+
```
Comment (pgunn) [inline caiman/source_extraction/volpy/utils.py:281]: Why the blank lines?

File: caiman/source_extraction/volpy/volparams.py [hunk 9]
```diff
@@ -26,24 +26,25 @@ def __init__(self, fnames=None, fr=None, index=None, ROIs=None, weights=None, do
         }
 
         self.volspike = {
-            'doCrossVal': doCrossVal, # cross-validate to optimize regression regularization parameters
-            'doGlobalSubtract': doGlobalSubtract,
```
Comment (pgunn) [inline caiman/source_extraction/volpy/volparams.py:30]: It might be easier to read this dict if the keys are listed in order

File: caiman/source_extraction/volpy/volpy.py [hunk 10]
```diff
@@ -22,64 +23,89 @@ class VOLPY(object):
...
+                 sigmas=np.array([1, 1.5, 2]), n_iter=2, weight_update='ridge', 
+                 do_plot=True, do_cross_val=False, sub_freq=75, 
+                 method='spikepursuit', superfactor=10, params=None):
+        
         """
+        Args:
             n_processes: int
-                number of processed used (if in parallel this controls memory usage)
-
+                number of processed used
```
Comment (pgunn) [inline caiman/source_extraction/volpy/volpy.py:36]: number of processes

File: caiman/source_extraction/volpy/volpy.py [hunk 11]
```diff
@@ -22,64 +23,89 @@ class VOLPY(object):
...
+                
+            flip_signal: boolean
+                whether to flip signal upside down for spike detection 
+                True for voltron, False for others
 
+            hp_freq_pb: float
+                high-pass frequency for removing photobleaching    
+            
             nPC_bg: int
                 number of principle components used for background subtraction
```
Comment (pgunn) [inline caiman/source_extraction/volpy/volpy.py:56]: principal

File: caiman/source_extraction/volpy/volpy.py [hunk 12]
```diff
@@ -22,64 +23,89 @@ class VOLPY(object):
...
+                'adaptive_threshold' method threshold based on estimated peak distribution
+                
+            min_spikes: int
+                minimal number of spikes to be detected
 
-            sigmas: 1-d array
-                spatial smoothing radius imposed on spatial filter
+            threshold: float
+                threshold for spike detection in 'simple' threshold method 
+                The real threshold is the value multiply estimated noise level
```
Comment (pgunn) [inline caiman/source_extraction/volpy/volpy.py:74]: multiplied by the estimated noise level?

File: caiman/source_extraction/volpy/volpy.py [hunk 13]
```diff
@@ -103,42 +128,42 @@ def fit(self, n_processes=None, dview=None):
...
+        self.estimates['num_spikes'] = np.array([results[i]['num_spikes'] for i in range(N)])
+        self.estimates['templates'] = np.array([results[i]['templates'] for i in range(N)])
+        self.estimates['snr'] = np.array([results[i]['snr'] for i in range(N)])
+        self.estimates['thresh'] = np.array([results[i]['thresh'] for i in range(N)])
+        self.estimates['spatial_filter'] = np.array([results[i]['spatial_filter'] for i in range(N)])
+        self.estimates['weights'] = np.array([results[i]['weights'] for i in range(N)])
+        self.estimates['locality'] = np.array([results[i]['locality'] for i in range(N)])
+        self.estimates['context_coord'] = np.array([results[i]['context_coord'] for i in range(N)])
+        self.estimates['F0'] = np.array([results[i]['F0'] for i in range(N)])
+        self.estimates['dFF'] = np.array([results[i]['dFF'] for i in range(N)])
```
Comment (pgunn) [inline caiman/source_extraction/volpy/volpy.py:166]: You could probably iterate over all of these more elegantly given that the keynames are the same

File: caiman/summary_images.py [hunk 14]
```diff
@@ -776,8 +780,12 @@ def local_correlations_movie_offline(file_name,
 
 
 def local_correlations_movie_parallel(params: Tuple) -> np.ndarray:
-    mv_name, idx, eight_neighbours, swap_dim, order_mean, ismulticolor, remove_baseline, winSize_baseline, quantil_min_baseline  = params
+    mv_name, idx, eight_neighbours, swap_dim, order_mean, ismulticolor, remove_baseline, winSize_baseline, quantil_min_baseline, gaussian_blur = params
     mv = cm.load(mv_name, subindices=idx, in_memory=True)
+    if gaussian_blur:
+        mv = mv.gaussian_blur_2D()
+    else:
+        pass
```
Comment (pgunn) [inline caiman/summary_images.py:788]: I don't think this else clause does anything

File: caiman/source_extraction/volpy/atm.py (hunk 2 above)
Comment (caichangjia) [reply caiman/source_extraction/volpy/atm.py:378]: Sorry this is my mistake. atm.py file should not be changed.

File: caiman/source_extraction/volpy/utils.py (hunk 6 above)
Comment (caichangjia) [reply caiman/source_extraction/volpy/utils.py:120]: What I think is the following: if they use Mask R-CNN, then they will install keras to run these lines. If they don't use Mask R-CNN, they can skip these steps and no need to install keras.

Comment (pgunn) [reply caiman/source_extraction/volpy/utils.py:120]: That's a fair point.

File: caiman/source_extraction/volpy/volparams.py (hunk 9 above)
Comment (caichangjia) [reply caiman/source_extraction/volpy/volparams.py:30]: They are listed in order not alphabetically, but more follow the sequence of their appearance in the algorithm.

File: caiman/source_extraction/volpy/utils.py (hunk 5 above)
Comment (caichangjia) [reply caiman/source_extraction/volpy/utils.py:59]: What I know is it works on Spyder. It serves the function to jump out a window (matplotlib auto) for user manually selecting neurons. Is there a better choice?

Comment (pgunn) [reply caiman/source_extraction/volpy/utils.py:59]: Spyder uses ipython as its background. You might want to test this code from the regular python commandline. I think one of the existing demos demonstrates how to do this with regular matplotlib code.

Comment (pgunn) [reply caiman/source_extraction/volpy/utils.py:59]: (I think this is the last thing that needs some attention in this diff; otherwise looking good)

Comment (pgunn) [reply caiman/source_extraction/volpy/utils.py]: looking good now

**Comment by pgunn** [comment]
Once this goes green, I'll merge (unless @epnev wants to chime in)

**Comment by epnev** [comment]
Just make sure that `demos/general/demo_pipeline_voltage_imaging.py` runs on a fresh environment when the code uses mask-RCNN. Right now this option is not activated @caichangjia

**Comment by caichangjia** [comment]
Yeah. I have tried yesterday on a fresh environment. Works well after installing keras manually.
